### PR TITLE
Task Based Flush API - FlushAsync - New Implementation

### PR DESCRIPTION
**Pull request by rajkumar-rangaraj**
Addressing #1743 

## FlushAsync

`FlushAsync` is a task based flush API for sending telemetry to application insights. 

### API
```
/// <summary>
/// Represents a task based flush for sending telemetry to application insights.
/// </summary>
public interface IAsyncFlushable : IDisposable
{
               /// <summary>
               /// Flushes the in-memory buffer asynchronously.
               /// </summary>
               /// <param name="cancellationToken">CancellationToken.</param>
               /// <returns>
               /// Returns true when telemetry data is transferred out of process (application insights server or local storage) and are emitted before the flush invocation.
               /// Returns false when transfer of telemetry data to server has failed with non-retriable http status code.
               /// </returns>
               Task<bool> FlushAsync(CancellationToken cancellationToken);
}
```

### Sample

```
class Program
{
    static void Main(string[] args)
    {
        TelemetryConfiguration configuration = TelemetryConfiguration.CreateDefault();
        configuration.InstrumentationKey = "IKEY";
        var telemetryClient = new TelemetryClient(configuration);
        telemetryClient.TrackTrace("Hello World!");

        // before exit, flush the remaining data
        var task = telemetryClient.FlushAsync(CancellationToken.None);

        try
        {
             await task;
        }
        catch(Exception ex)
        {
             throw ex;
        }
    }
}
```
### What is a transmission?

Telemetry items are batched at fixed interval (default is 30 secs) or whenever buffer is full (default is 500 items) or through manual flush. Batched items are Json serialized, compressed and converted into transmissions. These transmissions are sent to server. 

### Features
- Waits until all transmissions emitted before the flush invocation to complete
- Moves transmissions from buffer to disk
- On exception or throttle, moves transmission to disk
- Allow caller to decide when to cancel the method through `CancellationToken`

> :warning: &nbsp; Ideally, `Flush` or `FlushAsync` method should be used in the shutdown activity of an application.

## Scenarios
|Scenario| Functionality|
| ------------- | ------------- |
| Successful Transmission | <ul><li>Waits for all transmissions emitted before the flush invocation to complete</li><li>Moves items from local process buffer to disk (StorageFolder) on successful transmission</li><li>Marks `FlushAsync` task as success</li></ul>|
| Partial Transmission | <ul><li>Waits for all transmissions emitted before the flush invocation to complete</li><li> Creates new transmission from failed telemetry items within current transmission and moves it to disk</li><li> Moves local process buffer to disk</li><li>Marks `FlushAsync` task as success</li></ul>|
| Throttle | <ul><li> Transmission gets split into accepted and rejected transmission</li><li> Moves rejected transmission and local process buffer to disk</li><li> Waits for accepted transmission and all transmissions emitted before the flush invocation to complete</li><li>Marks `FlushAsync` task as success</li></ul>|
| Retriable Exception | <ul><li>Waits for all transmissions emitted before the flush invocation to complete</li><li> Moves current transmission and from local process buffer to disk</li><li>Marks `FlushAsync` task as success</li></ul>|
| Unknown Exception / Non-Whitelisted http status code | <ul><li>Waits for all transmissions emitted before the flush invocation to complete</li><li>Fails `FlushAsync` by sending return value as false</li></ul>|
| Storage issue | <ul><li>Fails `FlushAsync` by sending return value as false</li></ul>|

## Implementation

| InMemoryChannel | ServerTelemetryChannel |
| ------------- | ------------- |
| This channel offers minimal reliability guarantees and doesn't retry sending telemetry after a failure, hence `FlushAsync` method always returns true. [Learn More - Built-in telemetry channels]( https://docs.microsoft.com/en-us/azure/azure-monitor/app/telemetry-channels#built-in-telemetry-channels)   | A more advanced channel that has retry policies and the capability to store data on a local disk. Return true when telemetry data is transferred out of process (application insights server or local storage). Above explained scenarios are applicable for to this channel. |

- Both `ServerTelemetryChannel` and `InMemoryChannel` implements `IAsyncFlushable` and has implementation for `FlushAsync` API
- `TelemetrySerializer` serializes telemetries on a new task and enqueues to `Transmitter`
- `Transmission` now includes `TaskCompletionSource`. On `FlushAsync` call, `Task` is created from `TaskCompletionSource`. This `Task` reference is passed back to the caller of `FlushAsync` to await for completion. 
- Policies are updated to complete the `Transmission’s` `TaskCompletionSource` with success or failure status

## Checklist
- [X] I ran Unit Tests locally.
- [X] CHANGELOG.md updated with one line description of the fix, and a link to the original issue if available.

For significant contributions please make sure you have completed the following items:

- [ ] Design discussion issue #
- [ ] Changes in public surface reviewed

The PR will trigger build, unit tests, and functional tests automatically. Please follow [these](https://github.com/Microsoft/ApplicationInsights-dotnet/blob/develop/.github/CONTRIBUTING.md) instructions to build and test locally.

### Notes for authors:
- FxCop and other analyzers will fail the build. To see these errors yourself, compile localy using the Release configuration.

### Notes for reviewers:

- We support [comment build triggers](https://docs.microsoft.com/azure/devops/pipelines/repos/github?view=azure-devops&tabs=yaml#comment-triggers)
  - `/AzurePipelines run` will queue all builds
  - `/AzurePipelines run <pipeline-name>` will queue a specific build

### CODE REVIEW

File: BASE/src/Microsoft.ApplicationInsights/Channel/Transmission.cs [hunk 1]
```diff
@@ -21,9 +22,17 @@ public class Transmission
...
         private static HttpClient client = new HttpClient() { Timeout = System.Threading.Timeout.InfiniteTimeSpan };
-                
+        private static long flushAsyncCounter = 0;
+
         private int isSending;
 
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Transmission"/> class. This overload seperates telemetryitems serialization from object construction. 
+        /// </summary>
+        public Transmission()
```
Comment (reyang) [inline BASE/src/Microsoft.ApplicationInsights/Channel/Transmission.cs:32]: If the goal is to allow other packages (e.g. telemetry channel implementations) to access this ctor, a safer approach would be https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices.internalsvisibletoattribute.

**Review by reyang — APPROVED** [review]
LGTM with a non-blocking suggestion.

File: BASE/src/Microsoft.ApplicationInsights/Channel/Transmission.cs (hunk 1 above)
Comment (rajkumar-rangaraj) [reply BASE/src/Microsoft.ApplicationInsights/Channel/Transmission.cs:32]: @cijothomas, We had offline discussion about this topic and decided not to use internals visible. Introducing internals visible will limit our public API changes, should we reconsider  `InternalsVisibleToAttribute`?

Comment (cijothomas) [reply BASE/src/Microsoft.ApplicationInsights/Channel/Transmission.cs:32]: We can make this new fields internal, and use InternalsVisible approach now given that its intended to be accessible from ServerTelemetryChannel now.
If a need arises in the future, we can make it public, so other channels may leverage this.

File: BASE/src/Microsoft.ApplicationInsights/TelemetryClient.cs [hunk 2]
```diff
@@ -676,6 +678,38 @@ public void Flush()
...
+        /// <returns>
+        /// Returns true when telemetry data is transferred out of process (application insights server or local storage) and are emitted before the flush invocation.
+        /// Returns false when transfer of telemetry data to server has failed with non-retriable http status.
+        /// FlushAsync on InMemoryChannel always returns true, as the channel offers minimal reliability guarantees and doesn't retry sending telemetry after a failure.
+        /// </returns>
+        public Task<bool> FlushAsync(CancellationToken cancellationToken)
+        {
+            if (this.TryGetMetricManager(out MetricManager privateMetricManager))
+            {
+                privateMetricManager.Flush(flushDownstreamPipeline: false);
```
Comment (cijothomas) [inline BASE/src/Microsoft.ApplicationInsights/TelemetryClient.cs:693]: shouldn't we pass the cancellationToken to metricmanager.Flush() as well?

File: BASE/src/Microsoft.ApplicationInsights/Channel/Transmission.cs [hunk 3]
```diff
@@ -139,6 +141,32 @@ public ICollection<ITelemetry> TelemetryItems
             get; private set;
         }
 
+        /// <summary>
+        /// Gets the flush async id for the transmission.
+        /// </summary>
+        /// [EditorBrowsable(EditorBrowsableState.Never)]
```
Comment (TimothyMothra) [inline BASE/src/Microsoft.ApplicationInsights/Channel/Transmission.cs:147]: was this meant to be non-browsable?

File: BASE/src/Microsoft.ApplicationInsights/Channel/Transmission.cs [hunk 4]
```diff
@@ -139,6 +141,32 @@ public ICollection<ITelemetry> TelemetryItems
...
+        public long FlushAsyncId { get; } = Interlocked.Increment(ref flushAsyncCounter);
+
+        /// <summary>
+        /// Gets or sets a value indicating whether this transmission is created from FlushAsync API call.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public bool HasFlushTask { get; set; } = false;
```
Comment (cijothomas) [inline BASE/src/Microsoft.ApplicationInsights/Channel/Transmission.cs:154]: nit:
The name doesn't match the summary.. `IsCreatedFromFlush` matches the summary.

(Or is the summary actually incorrect?)

File: BASE/src/ServerTelemetryChannel/Implementation/TelemetryBuffer.cs [hunk 5]
```diff
@@ -205,6 +208,26 @@ IEnumerator IEnumerable.GetEnumerator()
             return this.GetEnumerator();
         }
 
+        private List<ITelemetry> MoveTelemeteryBufferForFlush()
```
Comment (cijothomas) [inline BASE/src/ServerTelemetryChannel/Implementation/TelemetryBuffer.cs:211]:
```suggestion
        private List<ITelemetry> GetBufferTelemetryAndResetBuffer()
```

nit.

File: BASE/src/ServerTelemetryChannel/Implementation/TelemetrySerializer.cs [hunk 6]
```diff
@@ -48,9 +50,35 @@ public virtual void Serialize(ICollection<ITelemetry> items)
...
+            var transmission = new Transmission(this.EndpointAddress, items)
+                                    { TransmissionStatusEvent = this.TransmissionStatusEvent };
             this.Transmitter.Enqueue(transmission);
         }
+
+        public virtual Task<bool> SerializeAsync(ICollection<ITelemetry> items, CancellationToken cancellationToken)
+        {
+            if (this.EndpointAddress == null)
+            {
+                throw new Exception("TelemetrySerializer.EndpointAddress was not set.");
```
Comment (cijothomas) [inline BASE/src/ServerTelemetryChannel/Implementation/TelemetrySerializer.cs:62]: would this get propagated to customer FlushAsync or we handle it somewhere before throwing up to user?

File: BASE/src/Microsoft.ApplicationInsights/Channel/Transmission.cs [hunk 7]
```diff
@@ -139,6 +141,32 @@ public ICollection<ITelemetry> TelemetryItems
...
+
+        /// <summary>
+        /// Serializes telemetry items.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public void Serialize(Uri address, IEnumerable<ITelemetry> telemetryItems, TimeSpan timeout = default(TimeSpan))
```
Comment (cijothomas) [inline BASE/src/Microsoft.ApplicationInsights/Channel/Transmission.cs:160]: minor point. I could be wrong as well.

feels a bit awkward for a instance method Serialize accept a list of items to be serialized...
Could we store the items in the ctor() itself, and Serialize() call simply serializes the items it already has?

File: BASE/src/ServerTelemetryChannel/TelemetryChannel.csproj [hunk 8]
```diff
@@ -44,5 +44,9 @@
     <PackageReference Include="System.IO.FileSystem.AccessControl" Version="4.7.0" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Compile Include="..\Microsoft.ApplicationInsights\Channel\TaskEx.cs" Link="TaskEx.cs" />
```
Comment (TimothyMothra) [inline BASE/src/ServerTelemetryChannel/TelemetryChannel.csproj:48]: please add this to the Common project instead. Line 51.

File: CHANGELOG.md [hunk 9]
```diff
@@ -68,6 +68,7 @@
 - [Fix: Disabling HeartBeats in Asp.Net Core projects causes Error traces every heart beat interval (15 minutes defualt)](https://github.com/microsoft/ApplicationInsights-dotnet/issues/1681)
 - [Standard Metric extractor (Exception,Trace) populates all standard dimensions.](https://github.com/microsoft/ApplicationInsights-dotnet/issues/1738)
 - [Add an explicit reference to System.Memory v4.5.4. This fixes a bug in System.Diagnostics.DiagnosticSource. We will remove this dependency when DiagnosticSource is re-released.](https://github.com/microsoft/ApplicationInsights-dotnet/issues/1707)
+- [New Task Based Flush API - FlushAsync] (https://github.com/microsoft/ApplicationInsights-dotnet/issues/1743)
```
Comment (TimothyMothra) [inline CHANGELOG.md:71]: this needs to be under the VNext section

File: BASE/src/ServerTelemetryChannel/Implementation/Transmitter.cs [hunk 10]
```diff
@@ -169,10 +173,84 @@ internal virtual void Enqueue(Transmission transmission)
 
             if (!this.Storage.Enqueue(transmissionGetter))
             {
+                transmission.HasFlushTask = false;
```
Comment (cijothomas) [inline BASE/src/ServerTelemetryChannel/Implementation/Transmitter.cs:176]: Trying to understand why does this has to be set to false here... Transmitter receives a transmission..How can it determine if HasFlushTask should be true or not, as HasFlushTask depends on if it was called from Flush..

File: BASE/src/ServerTelemetryChannel/Implementation/Transmitter.cs [hunk 11]
```diff
@@ -169,10 +173,84 @@ internal virtual void Enqueue(Transmission transmission)
...
+            var senderStatus = this.Sender.WaitForPreviousTransmissionsToComplete(cancellationToken).ConfigureAwait(false).GetAwaiter().GetResult();
+            if (senderStatus == TaskStatus.Canceled)
+            {
+                return TaskEx.FromCanceled<bool>(cancellationToken);
+            }
+
+            return senderStatus == TaskStatus.RanToCompletion && isStorageEnqueueSuccess ? this.successTask : this.failedTask; 
+        }
+
+        internal TaskStatus MoveTransmissionsAndWaitForSender(long transmissionFlushAsyncId, CancellationToken cancellationToken)
```
Comment (cijothomas) [inline BASE/src/ServerTelemetryChannel/Implementation/Transmitter.cs:240]: these 2 methods can be merged/made to call other, to avoid copy-pasting.

File: BASE/src/ServerTelemetryChannel/Implementation/Transmitter.cs [hunk 12]
```diff
@@ -169,10 +173,84 @@ internal virtual void Enqueue(Transmission transmission)
...
+        }
+
+        internal Task<bool> MoveTransmissionsAndWaitForSender(CancellationToken cancellationToken)
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return TaskEx.FromCanceled<bool>(cancellationToken);
+            }
+
+            var isStorageEnqueueSuccess = MoveTransmissions(this.Buffer.Dequeue, this.Storage.Enqueue, this.Buffer.Size);
```
Comment (cijothomas) [inline BASE/src/ServerTelemetryChannel/Implementation/Transmitter.cs:229]: Not sure where do we draw the line.. but this method is also a potentially long running (if high items and slow disk..), So should we pass cancellationtoken to it as well?

File: BASE/src/ServerTelemetryChannel/Implementation/TransmissionSender.cs [hunk 13]
```diff
@@ -134,6 +140,41 @@ public virtual bool Enqueue(Func<Transmission> transmissionGetter)
             return enqueueSucceded;
         }
 
+        /// <summary>
+        /// Wait for inflight transmissions to complete. Honors the CancellationToken set by an application on IAsyncFlushable.FlushAsync.
+        /// </summary>
+        internal Task<TaskStatus> WaitForPreviousTransmissionsToComplete(CancellationToken cancellationToken)
+        {
+            var transmissionFlushAsyncId = this.inFlightTransmissions.LastOrDefault().Key;
```
Comment (cijothomas) [inline BASE/src/ServerTelemetryChannel/Implementation/TransmissionSender.cs:148]: what exactly does LastOrDefault does here? Gets the transmission with highest id? or the one which was added most recent?

File: BASE/src/ServerTelemetryChannel/Implementation/TransmissionSender.cs [hunk 14]
```diff
@@ -155,7 +197,9 @@ private async Task StartSending(Transmission transmission)
                 try
                 {
                     TelemetryChannelEventSource.Log.TransmissionSendStarted(acceptedTransmission.Id);
-                    responseContent = await acceptedTransmission.SendAsync().ConfigureAwait(false);
+                    transmissionTask = acceptedTransmission.SendAsync();
+                    this.inFlightTransmissions.TryAdd(transmission.FlushAsyncId, transmissionTask);
```
Comment (cijothomas) [inline BASE/src/ServerTelemetryChannel/Implementation/TransmissionSender.cs:201]: adding to the dictionary can be done *before* called SendAsync() to ensure its always added to dictionary, irrespective of any exceptions.

File: BASE/src/Microsoft.ApplicationInsights/Channel/Transmission.cs (hunk 3 above)
Comment (rajkumar-rangaraj) [reply BASE/src/Microsoft.ApplicationInsights/Channel/Transmission.cs:147]: Yes. We are marking this properties/methods as internal. I will get this attribute removed.

File: BASE/src/Microsoft.ApplicationInsights/Channel/Transmission.cs (hunk 7 above)
Comment (rajkumar-rangaraj) [reply BASE/src/Microsoft.ApplicationInsights/Channel/Transmission.cs:160]: This is tricky, as we already have a constructor which takes the same number of arguments and serializes it.

File: BASE/src/Microsoft.ApplicationInsights/Channel/Transmission.cs (hunk 1 above)
Comment (TimothyMothra) [reply BASE/src/Microsoft.ApplicationInsights/Channel/Transmission.cs:32]: Can we make the InternalsVisible change as a separate PR, please?

File: BASE/src/ServerTelemetryChannel/Implementation/Transmitter.cs (hunk 10 above)
Comment (rajkumar-rangaraj) [reply BASE/src/ServerTelemetryChannel/Implementation/Transmitter.cs:176]: If a move to storage fails, we set this flag. `HasFlushTask` will be honored only in case of FlushAsync call.
`Transmitter.FlushAsync` calls `Enqueue` to follow the normal process. During that attempt if a transmission object fails to move to storage that needs to be tracked hence `HasFlushTask` is used as a marker.

File: BASE/src/ServerTelemetryChannel/Implementation/Transmitter.cs (hunk 11 above)
Comment (rajkumar-rangaraj) [reply BASE/src/ServerTelemetryChannel/Implementation/Transmitter.cs:240]: Return type of both of these methods are different and they call `WaitForPreviousTransmissionsToComplete' method with different parameters. 

Reason for me to split these functions are to avoid handling TaskCanceledException in the `Transmitter.FlushAsync` call. Felt handling exception is costlier here.

File: BASE/src/ServerTelemetryChannel/Implementation/Transmitter.cs (hunk 12 above)
Comment (rajkumar-rangaraj) [reply BASE/src/ServerTelemetryChannel/Implementation/Transmitter.cs:229]: We have to check the if cancellationtoken within the loop, this will impact performance as everytime we loop we need to check the status of cancellationtoken. If we are fine with this overhead I will add cancellationtoken check.

File: BASE/src/ServerTelemetryChannel/Implementation/TransmissionSender.cs (hunk 13 above)
Comment (rajkumar-rangaraj) [reply BASE/src/ServerTelemetryChannel/Implementation/TransmissionSender.cs:148]: The one that got added most recent.
Most of the times, recent transmission should be the one with highest id.

File: BASE/src/ServerTelemetryChannel/Implementation/TransmissionSender.cs (hunk 14 above)
Comment (rajkumar-rangaraj) [reply BASE/src/ServerTelemetryChannel/Implementation/TransmissionSender.cs:201]: `transmissionTask` is generated by `SendAsync()`. Only awaiting on a task will throw an exception. Before awaiting a task, an entry is added to dictionary. So irrespective of exception it gets added to dictionary.

File: BASE/src/Microsoft.ApplicationInsights/Channel/TaskEx.cs [hunk 15]
```diff
@@ -0,0 +1,37 @@
+﻿namespace Microsoft.ApplicationInsights.Channel
```
Comment (TimothyMothra) [inline BASE/src/Microsoft.ApplicationInsights/Channel/TaskEx.cs:1]: I see that you want to use this in multiple projects.
I recommend moving this to the "Common.Base" project at "BASE\src\Common\Common\".
You could do this as a separate PR.

File: BASE/src/Microsoft.ApplicationInsights/TelemetryClient.cs (hunk 2 above)
Comment (rajkumar-rangaraj) [reply BASE/src/Microsoft.ApplicationInsights/TelemetryClient.cs:693]: This requires a change to metrics flush API, should we explore it in a different PR?

File: BASE/src/Microsoft.ApplicationInsights/Channel/Transmission.cs (hunk 1 above)
Comment (cijothomas) [reply BASE/src/Microsoft.ApplicationInsights/Channel/Transmission.cs:32]: If its part of this PR, it'll simply the .public API file in this PR a lot.

Comment (rajkumar-rangaraj) [reply BASE/src/Microsoft.ApplicationInsights/Channel/Transmission.cs:32]: @TimothyMothra I agree with you. This has to be done separately. I will send a new PR and get that merged before this one.

File: BASE/src/ServerTelemetryChannel/Implementation/TelemetrySerializer.cs (hunk 6 above)
Comment (rajkumar-rangaraj) [reply BASE/src/ServerTelemetryChannel/Implementation/TelemetrySerializer.cs:62]: This gets propagated to customers FlushAsync. I had this test case to cover it.
```
            [TestMethod]
            public void ThrowsExceptionWhenEndpointAddressIsNull()
            {
                var serializer = new TelemetrySerializer(new StubTransmitter());
                AssertEx.Throws<Exception>(() => serializer.SerializeAsync(null, default));
            }
```
Now adding one more in TelemetryBufferTests to avoid confusion.

File: BASE/src/Microsoft.ApplicationInsights/Channel/TaskEx.cs (hunk 15 above)
Comment (rajkumar-rangaraj) [reply BASE/src/Microsoft.ApplicationInsights/Channel/TaskEx.cs:1]: We could completely remove common project with internals visible, working this on a separate PR.

File: BASE/src/Microsoft.ApplicationInsights/Channel/Transmission.cs (hunk 4 above)
Comment (rajkumar-rangaraj) [reply BASE/src/Microsoft.ApplicationInsights/Channel/Transmission.cs]: Apt name would be `IsFlushAsyncInProgress`. Changed both summary and property name.

File: BASE/src/ServerTelemetryChannel/Implementation/Transmitter.cs (hunk 12 above)
Comment (rajkumar-rangaraj) [reply BASE/src/ServerTelemetryChannel/Implementation/Transmitter.cs]: As per our offline discussion, code has to check IsCancellationRequested in every logical step. Hence added cancelationtoken check.

File: BASE/src/Microsoft.ApplicationInsights/TelemetryClient.cs (hunk 2 above)
Comment (cijothomas) [reply BASE/src/Microsoft.ApplicationInsights/TelemetryClient.cs:693]: please put a todo marker here.

File: BASE/src/Microsoft.ApplicationInsights/TelemetryClient.cs [hunk 16]
```diff
@@ -676,6 +678,38 @@ public void Flush()
...
+            if (pipeline != null)
+            {
+                MetricManager sharedMetricManager = pipeline.GetMetricManager(createIfNotExists: false);
+                sharedMetricManager?.Flush(flushDownstreamPipeline: false);
+
+                ITelemetryChannel channel = pipeline.TelemetryChannel;
+
+                if (channel is IAsyncFlushable asyncFlushableChannel && !cancellationToken.IsCancellationRequested)
+                {
+                    return asyncFlushableChannel.FlushAsync(cancellationToken);
```
Comment (cijothomas) [inline BASE/src/Microsoft.ApplicationInsights/TelemetryClient.cs:706]: Not really part of this PR, but we actually need to flush channels in *all* the sinks, as opposed to the default sink.
Please mark a TODO and we'll explore it after.

File: BASE/src/Microsoft.ApplicationInsights/Channel/InMemoryChannel.cs [hunk 17]
```diff
@@ -186,6 +188,25 @@ public void Flush(TimeSpan timeout)
...
 
+        /// <summary>
+        /// Will send all the telemetry items stored in the memory asynchronously.
+        /// </summary>
+        /// <param name="cancellationToken">CancellationToken.</param>
+        /// <returns>
+        /// This method is hard-coded to return true. This channel offers minimal reliability guarantees and doesn't retry sending telemetry after a failure.
+        /// </returns>
+        /// <remarks>
+        /// <a href="https://docs.microsoft.com/en-us/azure/azure-monitor/app/telemetry-channels#built-in-telemetry-channels">Learn more</a>
```
Comment (cijothomas) [inline BASE/src/Microsoft.ApplicationInsights/Channel/InMemoryChannel.cs:199]:
```suggestion
        /// <a href="https://docs.microsoft.com/azure/azure-monitor/app/telemetry-channels#built-in-telemetry-channels">Learn more</a>
```

File: BASE/src/Microsoft.ApplicationInsights/Channel/Transmission.cs [hunk 18]
```diff
@@ -358,6 +370,19 @@ public virtual async Task<HttpWebResponseWrapper> SendAsync()
             return Tuple.Create(transmissionA, transmissionB);
         }
 
+        /// <summary>
+        /// Serializes telemetry items.
+        /// </summary>
+        internal void Serialize(Uri address, IEnumerable<ITelemetry> telemetryItems, TimeSpan timeout = default(TimeSpan))
+        {
+            this.EndpointAddress = address;
```
Comment (cijothomas) [inline BASE/src/Microsoft.ApplicationInsights/Channel/Transmission.cs:378]: might be good to refactor (in a separate PR) this method, as this is doing more than serialization. (all the validations which were previously in ctor() is now done inside this method..)

File: BASE/src/ServerTelemetryChannel/Implementation/Transmitter.cs [hunk 19]
```diff
@@ -169,10 +173,84 @@ internal virtual void Enqueue(Transmission transmission)
...
+        {
+            TaskStatus taskStatus = TaskStatus.Canceled;
+            if (!cancellationToken.IsCancellationRequested)
+            {
+                transmission.IsFlushAsyncInProgress = true;
+                this.Enqueue(transmission);
+
+                try
+                {
+                    this.Storage.IncrementFlushAsyncCounter();
```
Comment (cijothomas) [inline BASE/src/ServerTelemetryChannel/Implementation/Transmitter.cs:191]: i think this is done to prevent items being moved from Storage to Buffer/Sender, while FlushAsync is in progress... If thats the case, then this must also be done in the `MoveTransmissionsAndWaitForSender(cancellationToken)` path as well.

**Review by cijothomas — APPROVED** [review]
left few comments. We can address them in follow ups.
Also - please move usage examples to this PR's description.

File: BASE/src/ServerTelemetryChannel/Implementation/Transmitter.cs [hunk 20]
```diff
@@ -169,10 +173,84 @@ internal virtual void Enqueue(Transmission transmission)
...
             }
         }
 
+        internal Task<bool> FlushAsync(Transmission transmission, CancellationToken cancellationToken)
+        {
+            TaskStatus taskStatus = TaskStatus.Canceled;
+            if (!cancellationToken.IsCancellationRequested)
+            {
+                transmission.IsFlushAsyncInProgress = true;
+                this.Enqueue(transmission);
```
Comment (cijothomas) [inline BASE/src/ServerTelemetryChannel/Implementation/Transmitter.cs:187]: There is a possibility that sender and buffer and storage are all full and data is dropped inside this method. We are blind to that possibility, and won't return false to end user in this case..

Comment (cijothomas) [reply BASE/src/ServerTelemetryChannel/Implementation/Transmitter.cs:187]: we can modify Enqueue to return its success status, and act acordingly. (in a future PR)

Comment (rajkumar-rangaraj) [reply BASE/src/ServerTelemetryChannel/Implementation/Transmitter.cs:187]: This use case is handled, if sender, buffer and storage are full IsFlushAsyncInProgress is set to false in Enqueue on line 176 `transmission.IsFlushAsyncInProgress = false;`

Line 210, in method FlushAsync there is a check to handle it.

File: BASE/src/Microsoft.ApplicationInsights/Channel/Transmission.cs (hunk 18 above)
Comment (rajkumar-rangaraj) [reply BASE/src/Microsoft.ApplicationInsights/Channel/Transmission.cs:378]: Added a TODO in summary.
